tests/torch_trainer.py

```diff
@@ -0,0 +1,148 @@
+                    # do forward pass
+                    bufs, out = module_func(params, bufs, inputs)
+
+                    # do some debugging when in local mode
+                    if alpa.torch.mode() == "local":
+                        print("out: ", out)
+
+                    # do loss computation
+                    loss_value = loss_func(out, targets)
+                    return loss_value, bufs
+
+                # do model forward + backward pass
+                params_grad, (loss_value, bufs) = atorch.grad_and_value(compute_loss, has_aux=True)(params, bufs, inputs, targets)
+                # (loss_value, bufs), params_grad = alpa.value_and_grad(compute_loss, has_aux=True)(params, bufs, inputs, targets)
+
+                # do optimizer step
+                params, optim_state = optim_func(params, params_grad, optim_state)
+
+                return params, bufs, optim_state, loss_value
+
+            train_func = functools.partial(sgd_train_func, module_func, loss_func, optim_func)
+            parallel_method = alpa.PipeshardParallel(stage_mode="auto", num_micro_batches=num_micro_batches)
+            # parallel_method = alpa.PipeshardParallel(stage_mode="auto", num_micro_batches=num_micro_batches, submesh_physical_shape_space="all", submesh_logical_shape_space="all")
+            # parallel_method = alpa.PipeshardParallel(stage_mode="uniform", num_micro_batches=num_micro_batches)
+            # parallel_method = alpa.ShardParallel(num_micro_batches=num_micro_batches)
+            # parallel_method = alpa.ShardParallel()
+
+            def iter_func(params, bufs, optim_state, pt_batch):
+                if alpa.torch.mode() == "local":  # local development
+                    params, bufs, optim_state, loss_value = train_func(
+                        params,
+                        bufs,
+                        optim_state,
+                        pt_batch,
+                    )
+                    # Show that outputs are in PT format (no need in actual training code)
+                    alpa.torch.assert_format("torch", params, bufs, optim_state, loss_value)
+                elif alpa.torch.mode() == "dist":  # distributed training
+                    params, bufs, optim_state, loss_value = alpa.parallelize(
+                        train_func,
+                        method=parallel_method,
+                        batch_argnums=(3,),  # NOTE: assumes the 4th argument is input batch
+                        donate_argnums=(0, 1, 2),  # NOTE: preserves memory addr and sharding spec for first 3 args
+                    )(
+                        params,
+                        bufs,
+                        optim_state,
+                        alpa.torch.to_format("alpa", pt_batch),
+                    )
+                    # Show that outputs are in Alpa format (no need in actual training code)
+                    alpa.torch.assert_format("alpa", params, bufs, optim_state, loss_value)
```

Comment:
It is better to call `alpa.parallelize` outside of the loop. This is more friendly for caching and reduces the overhead (although it is very small currently). 

```suggestion
            if alpa.torch.mode() == "dist":
                train_func = alpa.parallelize(
                        train_func,
                        method=parallel_method,
                        batch_argnums=(3,),  # NOTE: assumes the 4th argument is input batch
                        donate_argnums=(0, 1, 2),  # NOTE: preserves memory addr and sharding spec for first 3 args
                    )
               
            def iter_func(params, bufs, optim_state, pt_batch):
                    params, bufs, optim_state, loss_value = train_func(
                        params,
                        bufs,
                        optim_state,
                        alpa.torch.to_format(mode, pt_batch),
                    )
                    alpa.torch.assert_format(mode, params, bufs, optim_state, loss_value)
```